### ISSUE TITLE
I'm trying to determine the reason for Code 10, Connection lost issues

### ISSUE THREAD

**Issue by oferchen** [ronf/asyncssh#266]
```
async def run_client_retry(host, command):
	attempts = 0
	timeout = 5
	conn = None
	command = 'echo "hello world"'
	while attempts < 5:
		try:
			conn = await asyncio.wait_for(asyncssh.connect(host, username=username), timeout=timeout)
			break
		except (OSError, asyncssh.Error) as e:
			print('Connection has caught an exception - {}'.format(str(e.__dict__)))
			attempts += 1
	if conn:
		return await conn.run(command, check=True)
	else:
		print('Connection has an issue - {}'.format(str(conn)))

async def getsshdata(d={}):
	tasks = []
	tasks.append( (run_client_retry(host, command)) )
	results = await asyncio.gather(*tasks, return_exceptions=True)
	for i, result in enumerate(results, 1):
		if isinstance(result, Exception):
			print('Task {} failed: {}'.format(i, str(result)))
			print(result.__dict__)
		elif result.exit_status != 0:
			print('Task {} exited with status {}:'.format(i, result.exit_status))
			print(result.stderr, end='')
		else:
			print('Task {} succeeded:'.format(i))
```

I always get several connection lost errors, it's so consistent that I think there is another issue (it might just be ssh server max connections).
I need some assistance to debug the issue or any ideas you might have towards resolving it.

**Comment by ronf** [comment]
Sorry for the slow response. The e-mail notification about this was incorrectly flagged as spam. Are you still running into this issue?

There are limits on most SSH servers on how many connections you can open at once. With OpenSSH, I think the default is no more than 10 connections which can be going through SSH authentication in parallel. So, if you want more connections than that, you have to space them out to allow the authentication to complete on existing connections before opening additional ones. You can raise this limit if you have access to the server's config -- I think the config setting is named "MaxStartups". There's also a limit on the number of allowed sessions on a single connection, controlled by "MaxSessions", but you seem to be opening a new connection each time here, so the latter setting shouldn't be an issue.

You may want to enable debugging to see exactly how far it gets before it fails, but if you're seeing a mix of successes and failures when opening many connections in parallel and it is reporting back "Connection lost", I think the "MaxStartups" is most likely your issue.

To enable debug logging, you'd do something like:

```python
    import logging
    logging.basicConfig(level='DEBUG')
    asyncssh.set_debug_level(2)
```

**Comment by oferchen** [comment]
it appears that when I retry the connection it always go through so I don't think this situation matches the MaxSessions limit
the above snippet did resolve most of the issues for me (as long as the connections are coming through)

**Comment by ronf** [comment]
The limit is on the number of simultaneous sessions which haven't finished authenticating yet. So, retrying would be enough to fix things in some cases, if other connections you had open managed to finish authenticating before you retried.

**Comment by oferchen** [comment]
it would be a bit faster if the retry value would be included in asyncssh but I guess I could override that method myself.
Thanks for your help and your hard work!

**Comment by ronf** [comment]
If the source of the error is related to MaxStartups, you'd actually be better off using an asyncio.Semaphore() to limit the number of simultaneous calls to asyncssh.connect() rather than overwhelming the server and having it react by failing some of the connections. It's actually quite simple to do this. You'd just create a Semaphore() with a limit of the max number of simultaneous connects you want to allow and grab the semaphore while you are calling asyncssh.connect(). For instance:

```python
class MySSHClient:
    def __init__(self, max_connects):
        self._semaphore = asyncio.Semaphore(max_connects)

    async def connect(self, host):
        async with self._semaphore:
            return await asyncssh.connect(host)
        
    async def run_command(self, host, command):
        conn = await self.connect(host)

        async with conn:
            return await conn.run(command)

async def run():
    client = MySSHClient(5)
    command = 'echo "hello world"'
    
    tasks = [client.run_command(host, command) for host in 100*['localhost']]

    for task in asyncio.as_completed(tasks):
        result = await task

        if isinstance(result, Exception):
            print(f'Task failed: {result}')
            print(result.__dict__)
        elif result.exit_status != 0:
            print(f'Task exited with status {result.exit_status}')
            print(result.stderr, end='')
        else:
            print(f'Task succeeded:')
            print(result.stdout, end='')
```

Basically, the only change is to initialize a _semaphore member in MySSHClient and then to do an "async with" on that in a wrapper around asyncssh.connect().

Note that this allows more than 5 SSH connections to be open at once, but it limits things such that no more than 5 connections can be performing the SSH handshake at once, avoiding hitting the server's MaxStartups limit.

This should be much friendlier to the server than opening a large number of connections all at once and having the server fail some of them, which then causes the client to open even more connections when it retries. Here, it should pace the connections to make sure that none of them fail in the first place.
